test/unit/ReservoirRouter.t.sol

```diff
@@ -0,0 +1,759 @@
+        (uint112 lReserve0, uint112 lReserve1, ) = _constantProductPair.getReserves();
+        uint256 lAmountOut = bound(aAmountOut, 1000, lReserve1 / 2);
+
+        // arrange
+        uint256 lSwapFee = _constantProductPair.swapFee();
+
+        // act
+        uint256 lAmountIn = _router.getAmountIn(lAmountOut, lReserve0, lReserve1, 0, lSwapFee, ExtraData(0,0,0));
+        _tokenA.mint(address(_constantProductPair), lAmountIn);
+        uint256 lActualAmountOut = _constantProductPair.swap(-int256(lAmountOut), false, address(this), bytes(""));
+
+        // assert
+        assertLt(lAmountOut, lAmountIn);
+        assertEq(lAmountOut, lActualAmountOut);
+    }
+
+    function testGetAmountIn_SP(uint256 aAmountOut) public
+    {
+        // assume
+        (uint112 lReserve0, uint112 lReserve1, ) = _stablePair.getReserves();
+        uint256 lAmountOut = bound(aAmountOut, 1, lReserve1 / 2);
+
+        // arrange
+        uint256 lSwapFee = _stablePair.swapFee();
+        uint64 lToken0PrecisionMultiplier = ReservoirLibrary.getPrecisionMultiplier(_stablePair.token0());
+        uint64 lToken1PrecisionMultiplier = ReservoirLibrary.getPrecisionMultiplier(_stablePair.token1());
+        uint64 lA = ReservoirLibrary.getAmplificationCoefficient(address(_stablePair));
+
+        // act
+        uint256 lAmountIn
+            = _router.getAmountIn(
+                lAmountOut,
+                lReserve0,
+                lReserve1,
+                1,
+                lSwapFee,
+                ExtraData(lToken0PrecisionMultiplier,lToken1PrecisionMultiplier, lA)
+        );
+        _tokenA.mint(address(_stablePair), lAmountIn);
+        uint256 lActualAmountOut = _stablePair.swap(-int256(lAmountOut), false, address(this), bytes(""));
+
+        // assert
+        assertLt(lAmountOut, lAmountIn);
+        assertEq(lAmountOut, lActualAmountOut);
+    }
+
+    function testGetAmountsOut_CP(uint256 aAmtBToMint, uint256 aAmtCToMint, uint256 aAmtIn) public
+        returns (uint256 lAmtIn, uint256[] memory lAmounts, address[] memory lPath, uint256[] memory lCurveIds)
+    {
+        // assume
+        uint256 lAmtBToMint = bound(aAmtBToMint, 2e3, type(uint112).max / 2);
+        uint256 lAmtCToMint = bound(aAmtCToMint, 2e3, type(uint112).max / 2);
+        lAmtIn = bound(aAmtIn, 2e3, type(uint112).max / 2);
+
+        // arrange
+        ConstantProductPair lOtherPair = ConstantProductPair(_createPair(address(_tokenB), address(_tokenC), 0));
+        _tokenB.mint(address(lOtherPair), lAmtBToMint);
+        _tokenC.mint(address(lOtherPair), lAmtCToMint);
+        lOtherPair.mint(address(this));
+
+        // act
+        lPath = new address[](3);
+        lPath[0] = address(_tokenA);
+        lPath[1] = address(_tokenB);
+        lPath[2] = address(_tokenC);
+        lCurveIds = new uint256[](2);
+        lCurveIds[0] = 0;
+        lCurveIds[1] = 0;
+
+        lAmounts = _router.getAmountsOut(lAmtIn, lPath, lCurveIds);
+
+        // assert
+
```

Comment:
Not sure how to fuzz outside of just implementing the x*y=k math again in a helper... Maybe just write some discrete cases? (also see my comment about moving the library tests into their own file).

Reply:
okay made the test cases discrete instead of fuzzed so we have a number to assert for, in [this commit](https://github.com/vexchange/v3-periphery/pull/1/commits/27e8fc2f767c45ac5dd0a95b6067cc85961bb257)